rpt.js

```diff
@@ -43,23 +44,24 @@ Node.prototype.path = ''
 Node.prototype.realpath = ''
 Node.prototype.children = null
 
-function Link (pkg, path, realpath, cache) {
-  if (cache[path]) return cache[path]
+function Link (pkg, logical, physical, realpath, cache) {
```

Comment:
Which is to say, the logical argument is the path as it appears if you are blind to symlinks.
The physical argument is the path if you resolve the symlinks in the root, but are blind to the symlinks in the module itself.
And finally, realpath is the full path with all symlinks resolved.

We have to keep logical around so that the paths in the data structure we return are inside of the root that we were given. Not having that confuses npm, as it expects to be able to check that kind of thing.

We have to keep physical around so that we actually know if the node is contained within a symlink.

And finally, realpath is compared to physical for that same symlink check.

An alternate strategy would be to change to lstating each module to see if it's a link, and maintaining the link status down the tree based on parent.isLink.

The advantage of the current strategy is that it only involves recursive readdirs.